doc/cwtool.1
.ds Q" ""
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..

.TH cwtool 1 "cw 0.13"

.SH NAME
cwtool \- tool for reading and writing floppy disk images with a Catweasel controller

.SH SYNOPSIS
.B cwtool
\-V

.B cwtool
\-D

.B cwtool
\-I
[\-v]
[\-n]
[\-f \fI<file>\fR]
[\-e \fI<config>\fR]

.B cwtool
\-L
[\-v]
[\-n]
[\-f \fI<file>\fR]
[\-e \fI<config>\fR]

.B cwtool
\-S
[\-v]
[\-n]
[\-f \fI<file>\fR]
[\-e \fI<config>\fR]
\fI<diskname>\fR
\fI<srcfile|device>\fR

.B cwtool
\-R
[\-v]
[\-n]
[\-f \fI<file>\fR]
[\-e \fI<config>\fR]
[\-r \fI<num>\fR]
[\-o \fI<file>\fR]
\fI<diskname>\fR
\fI<srcfile|device>\fR
[\fI<srcfile>\fR ...]
\fI<dstfile>\fR

.B cwtool
\-W
[\-v]
[\-n]
[\-f \fI<file>\fR]
[\-e \fI<config>\fR]
[\-s]
\fI<diskname>\fR
\fI<srcfile>\fR
\fI<dstfile|device>\fR

.SH DESCRIPTION
.PP
\fBcwtool\fR is the user space companion program for the cw kernel driver module. cw is a package for the Catweasel controller especially for accessing the floppy drives connected to Catweasel. Some preliminary remarks:
.RS
.IP * 2
\fBcwtool\fR uses one number for track numbering (you see those numbers with the options \-R \-v or \-W \-v). This number contains the hardware track and side. How the conversion is done depends on the configuration of the selected disk format. In most cases the track number is equal to 2 * hardware_track + side and is within the range from 0 to 171.
.IP * 2
Options like \-D, \-n, \-f or \-e refer to a configuration. The configuration syntax is currently undocumented, because it may change in future versions.
.IP * 2
If you want to use files named \- use ./\-, because otherwise \fBcwtool\fR would use stdin or stdout (depending on the file being source or destination).
.IP * 2
\fBcwtool\fR reads or writes disk images, it does not care if these images contain valid file systems or not. If you want to access the filesystem within an image, you have to use separate tools. Some tools are listed below.
.IP * 2
Not all formats have been extensively tested, if you have problems please contact the author (the email address is listed in the README of the source distribution).
.IP * 2
To allow reading and writing of G64 disk images \fBcwtool\fR has some built\-in magic. Before writing such images to disk unneeded data is stripped away, to shorten the track. Otherwise the track data may be too long and the last bytes of the track would overwrite the SYNC of the first sector. This does only work if the G64 image does contain valid CBM sector header ids, if not the data is written unchanged and may be too long. After reading a disk the same routines are used to strip away the track gap. It may be possible that too many data is discarded and the image is unusable (especially disks with copy protection mechanisms may use obscure sector layouts or other things). If you have such disks please send the author a raw image of it (look below how to create one). Additionally after reading the data after a SYNC is also byte aligned, because it seems VICE does not like unaligned data. To see what the track stripping and aligning routines are doing use \-v \-v \-v \-v.
.RE

.SH OPTIONS
.IP "\-V, \-\-version" 8
Print out version and exit.
.IP "\-D, \-\-dump" 8
Dump builtin config and exit.
.IP "\-I, \-\-initialize" 8
Initialize the configured devices.
.IP "\-L, \-\-list" 8
List available disk names and exit.
.IP "\-S, \-\-statistics" 8
Print out statistics of a disk, most notably the histogram.
.IP "\-R, \-\-read" 8
Read a disk and write the content to an image file. In combination with \-v a detailed report about bad sectors is given, the format is ss=ee@0xhhhhhh, with:
.RS
.IP * 2
ss \-\- sector number
.IP * 2
ee \-\- error description: nf \- not found, en \- encoding error (MFM, GCR, ...), nu \- wrong track/other number, si \- wrong sector size, cs \- checksum error, id \- wrong id, mx \- mixed error
.IP * 2
hhhhhh \-\- hexadecimal offset in destination file
.RE
.IP "\-W, \-\-write" 8
Write a disk with content read from an image file.
.IP "\-h, \-\-help" 8
Print out usage information.
.IP "\-v, \-\-verbose" 8
Increase verbosity level.
.IP "\-n, \-\-no\-rcfiles" 8
Do not read rcfiles (/etc/cwtoolrc and ~/.cwtoolrc) on startup.
.IP "\-f \fI<file>\fR, \-\-config \fI<file>\fR" 8
Read additional config from \fI<file>\fR.
.IP "\-e \fI<config>\fR, \-\-evaluate \fI<config>\fR" 8
Evaluate the given string \fI<config>\fR as configuration parameters.
.IP "\-r \fI<num>\fR, \-\-retry \fI<num>\fR" 8
Retry \fI<num>\fR times on read errors.
.IP "\-o \fI<file>\fR, \-\-output \fI<file>\fR" 8
output raw data of bad sectors to \fI<file>\fR.
.IP "\-s, \-\-ignore\-size" 8
Do not check if source file contains more or less bytes than needed.

.SH EXAMPLES
.IP "1." 8
.Vb
\&\fBcwtool\fR \-L \-v
.Ve
Get a list of all available disk names with some additional information.

.IP "2." 8
.Vb
\&\fBcwtool\fR \-S raw_14 /dev/cw0raw0
.Ve
Access first floppy drive on first controller and print out a histogram of a disk read with 14 MHz.

.IP "3." 8
.Vb
\&\fBcwtool\fR \-R \-v msdos_hd /dev/cw0raw0 bootdisk.img
.Ve
Read an MSDOS HD disk and write the content to file bootdisk.img. Print out some status information while reading.

.IP "4." 8
.Vb
\&\fBcwtool\fR \-W \-v c1541 image.d64 /dev/cw2raw1
.Ve
Write a C1541 disk with the content read from image.d64 to the second floppy drive of the third controller.

.IP "5." 8
.Vb
\&\fBgzip\fR \-d \-c image.adf.gz |
\&\fBcwtool\fR \-W \-v amiga_dd \- /dev/cw0raw0
.Ve
Uncompress image.adf.gz on the fly and write it to disk.

.IP "6." 8
.Vb
\&\fBcwtool\fR \-W \-v amiga_dd image.adf /dev/cw0raw0 &&
\&\fBcwtool\fR \-R amiga_dd /dev/cw0raw0 \- |
\&\fBcmp\fR \- image.adf
.Ve
Write image.adf to disk, read it in and compare with the original data, if there is a difference \fBcmp\fR will exit with an error message.

.IP "7." 8
.Vb
\&\fBcwtool\fR \-R \-r 0 \-v raw_14 /dev/cw0raw0 \- |
\&\fBbzip2\fR \-c \- > image.cwraw.bz2
.Ve
Read a disk with 14 MHz and compress the data on the fly. The written data is in raw format, this means it contains just the values \fBcwtool\fR got from the kernel driver. This is useful for later analysis of the disk format, if the format is currently not supported by \fBcwtool\fR. DD disks should be read with raw_14 and HD disks with raw_28. Giving higher values to \-r means that more copies of every track will be saved. This will enlarge the raw image files noticeable.

.IP "8." 8
.Vb
\&\fBcwtool\fR \-S \-v \-f \- my_raw /dev/cw0raw0 <<EOC
\&disk \*(Q"my_raw\*(Q"
\&        {
\&        format \*(Q"raw\*(Q"
\&        image \*(Q"raw\*(Q"
\&        clock 28
\&        track 120 timeout 500
\&        }
\&EOC
.Ve
Read track 120 (hardware track 60, first side) of a disk with 28 MHz, print out a long histogram.

.IP "9." 8
.Vb
\&\fBcwtool\fR \-W \-v msdos_hd /dev/zero /dev/cw0raw0
.Ve
Format an MSDOS HD disk with all sectors containing zeros.

.IP "10." 8
.Vb
\&\fBcwtool\fR \-W \-v amiga_dd /dev/urandom /dev/cw0raw0
.Ve
Format an Amiga DD disk with all sectors containing random data.

.IP "11." 8
.Vb
\&\fBcwtool\fR \-R \-r 0 \-v \-v \-v \-v \-v \-v amiga_dd  \\
\&        /dev/cw0raw0 /dev/null 2> log
.Ve
Write a more verbose report of what happened while reading an Amiga DD disk to log, do not retry on disk errors.

.IP "12." 8
.Vb
\&\fBcwtool\fR \-W \-v c1541_42_g64 image.g64 /dev/cw2raw1
.Ve
Write a C1541 disk with the content read from image.g64 to the second floppy drive of the third controller. If image.g64 contains more or less than the needed 42 tracks you will get warnings.

.IP "13." 8
.Vb
\&\fBcwtool\fR \-I \-f \- <<EOC
\&drive \*(Q"/dev/cw0raw0\*(Q"
\&        {
\&        ignore_diskchange yes
\&        double_step yes
\&        }
\&EOC
.Ve

NEWS
cw 0.13-20131228
	* kernel driver: fixed the Makefile to handle kernel version
	  strings like 3.8-trunk correctly
	* on non-linux platforms the driver build is now omitted and cwtool is
	  built standalone
	* included support to build cwtool together with the catweasel OS X
	  driver from http://catweasel-osx.sourceforge.net/ (thanks to
	  David Schmidt)

cw 0.13-20130306
	* man page: added information about the difference between 5.25" DD
	  and HD disks (thanks to Levente Harsfalvi)

cw 0.13-20121202
	* kernel driver: added support for the MK2 Anniversary Edition taken
	  from https://github.com/ezrec/cw/ (thanks to Jason McMullan).
	  This support requires a modified 40-pin IDE cable with pins 37
	  and 38 (IDE_CS0 and IDE_CS1) swapped and you need to use the
	  MK2_PORTS parameter when executing 'make install'. In addition,
	  make sure that other drivers do not use these ports 
	* added configuration mac_ssdd_400 and mac_ssdd_400_524 for Mac400
	  disks. while mac_ssdd_400_524 writes out all 524 bytes of a
	  sector, mac_ssdd_400 gives back 512 bytes per sector
	* added compile time fixes to build cwtool standalone on non-linux
	  platforms (thanks to balrog). This is experimental

cw 0.13-20120623
	* kernel driver: compile time fixes for Linux 3.0+ (thanks to
	  Dennis Boone)
	* kernel driver: compile time fixes for Linux 2.6.38+ (thanks to
	  Adrian Glaubitz)

cw 0.13-20090511
	* included cwio as interface to the kernel driver for other programs
	* included new FPGA firmware for MK4 cards
	* support for Victor 9000 disks was added (thanks to Jonathan
	  Gevaryahu for supplying raw images)
	* kernel driver: raw values to be written to disk are now subtracted
	  from 0x7f instead of 0x80
	* changed handling of raw images. raw images created with the current
	  version are not compatible with earlier versions of cwtool, but it
	  is still possible to read older raw images with the current cwtool
	* new option: -o
	* support for textual raw images (useful in combination with -o)
	* consolidated naming scheme for disk formats (thanks to Jonathan
	  Gevaryahu)
	* changed configuration for mac_dsdd_800 and mac_dsdd_800_524.
	  these formats are now using an experimental error correction.
	  this may reduce read errors in some situations (thanks to Yves
	  Benabou for supplying raw images)
	* when writing D64 images the disk id used in sector headers is now
	  taken from the BAM of the D64 image (offsets 0x165a2 and 0x165a3
	  in the D64 image)
	* corrected c1541 formats for 300 rpm drives with 40 and 42 tracks
	* drive directives in /etc/cwtoolrc and ~/.cwtoolrc are now also
	  used with cwtool -R and cwtool -W

cw 0.12.3-20080211
	* changed configuration mac_5.25. writing such images should now be
	  possible (thanks to Matt Carpenter for supplying raw images and
	  testing written images with a real Apple IIc)
	* added configuration mac_5.25_300 for Apple IIe 5.25" disks. this
	  format is an interims solution to support drives with 300 rpm.
	  but do not rely on the existence of these formats in future
	  versions, the handling of drives with different rpm's may change
	  again
	* compile time fixes for Linux 2.6.19+

cw 0.12.2-20070324
	* corrected track ranges in configurations for ibm_dsdd and dg_dsdd
	* lowered needed sync_length for various c1541 configurations
	* adjusted Makefile of the driver to fix compilation problems under
	  Linux 2.4

cw 0.12.1-20070319
	* slightly changed configuration for mac_dd_800 and mac_dd_800_524
	* enabled post compensation in all disk configurations. this should
	  improve readability of disks but will also slow down disk reading
	  especially on older computers with CPUs below 1 GHz

cw 0.12-20070304
	* support for Mac800K disks was added (thanks to Steve Ocepek,
	  Oliver Stabel and Stefan A. Haubenthal for supplying raw images;

src/cwtool/cwtool.c
/****************************************************************************
 ****************************************************************************
 *
 * cwtool.c
 *
 ****************************************************************************
 *
 * very short documentation
 *
 * - struct disk is the center of all things, this struct contains the
 *   layout of all supported disks.
 * - struct disk has an entry for each track (each track may have a
 *   different format)
 * - L0: catweasel raw counter values
 *   L1: raw bits (FM, MFM, GCR, ..., converted from/to L0 with bounds)
 *   L2: decoded FM, MFM, GCR, ... (with checksums and sector headers)
 *   L3: sector data bytes only
 * - if a disk should be read, disk_read() is called, it opens the given
 *   source and destination file, source file has to be L0 (either
 *   catweasel device or file containing raw counter values), destination
 *   file may be L0 - L3 depending on the selected format
 * - disk_write() is more or less inverse, source file may be L0 - L3,
 *   depending on selected format, destination file has to be L0
 * - plain image formats as ADF or D64 are L3, G64 is L1
 * - config.c translates a textual config file to entries in struct disk
 *   beside the built in directives it also understands format specific
 *   directives passed via struct format_desc
 * - struct fifo is the container for the data and some side information,
 *   its functions support bit and byte operations suitable for decoding
 *   and encoding
 *
 ****************************************************************************
 ****************************************************************************/





#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#include "cwtool.h"
#include "error.h"
#include "debug.h"
#include "verbose.h"
#include "config.h"
#include "disk.h"
#include "drive.h"
#include "file.h"
#include "string.h"
#include "version.h"



#define PROGRAM_NAME			"cwtool"
#define VERSION_STRING			PROGRAM_NAME " " VERSION "-" VERSION_DATE

#define CWTOOL_MAX_CONFIG_FILES		64
#define CWTOOL_MAX_CONFIG_EVALS		64
#define CWTOOL_MAX_CONFIG_SIZE		0x10000
#define CWTOOL_MODE_VERSION		1
#define CWTOOL_MODE_DUMP		2
#define CWTOOL_MODE_INITIALIZE		3
#define CWTOOL_MODE_LIST		4
#define CWTOOL_MODE_STATISTICS		5
#define CWTOOL_MODE_READ		6
#define CWTOOL_MODE_WRITE		7



char					program_name[] = PROGRAM_NAME;
static int				mode;
static char				*disk_name;
static char				*files[CWTOOL_MAX_IMAGES];
static int				files_count;
static int				no_rcfiles;
static char				*config_files[CWTOOL_MAX_CONFIG_FILES];
static int				config_files_count;
static char				*config_evals[CWTOOL_MAX_CONFIG_EVALS];
static int				config_evals_count;
static char				config_type[CWTOOL_MAX_CONFIG_FILES + CWTOOL_MAX_CONFIG_EVALS];
static int				config_types_count;
static int				retry = 5;
static int				ignore_size;
static FILE				*info_fh;



/****************************************************************************
 * do_snprintf2
 ****************************************************************************/
static int
do_snprintf2(
	int				len,
	int				max)

	{
	if ((len == -1) || (len >= max)) error_message("do_snprintf() size exceeded");
	return (len);
	}



/****************************************************************************
 * do_snprintf
 ****************************************************************************/
#define do_snprintf(string, max, args...)	do_snprintf2(snprintf(string, max, args), max)



/****************************************************************************
 * config_read_file
 ****************************************************************************/
static void
config_read_file(
	char				*path,
	int				flags)

	{
	struct file			fil;
	char				data[CWTOOL_MAX_CONFIG_SIZE];
	int				size;

	if (! file_open(&fil, path, FILE_MODE_READ, flags)) return;
	size = file_read(&fil, (unsigned char *) data, sizeof (data));
	if (size == sizeof (data)) error("file '%s' too large for config", path);
	file_close(&fil);
	config_parse(path, data, size);
	}



/****************************************************************************
 * config_read_rc_files
 ****************************************************************************/
static void
config_read_rc_files(
	void)

	{
	char				*home, path[CWTOOL_MAX_PATH_LEN];

	config_read_file("/etc/cwtoolrc", FILE_FLAG_RETURN);
	home = getenv("HOME");
	if (home == NULL) return;
	do_snprintf(path, sizeof (path), "%s/.cwtoolrc", home);
	config_read_file(path, FILE_FLAG_RETURN);
	}



/****************************************************************************
 * config_get
 ****************************************************************************/
static void
config_get(
	void)

	{
	char				number[128];
	int				c, e, f;

	config_parse("(builtin config)", config_default, string_length(config_default));
	if (! no_rcfiles) config_read_rc_files();
	for (c = e = f = 0; c < config_types_count; c++)
		{
		if (config_type[c] == 'f')
			{
			config_read_file(config_files[f++], FILE_FLAG_NONE);
			continue;
			}
		do_snprintf(number, sizeof (number), "-e/--evaluate #%d", e + 1);
		config_parse(number, config_evals[e], string_length(config_evals[e]));
		e++;
		}
	}



/****************************************************************************
 * cmd_version
 ****************************************************************************/
static void
cmd_version(
	void)

	{
	printf(VERSION_STRING "\n");
	}



/****************************************************************************
 * cmd_dump
 ****************************************************************************/
static void
cmd_dump(
	void)

	{
	printf("# builtin default config (" VERSION_STRING ")\n\n%s", config_default);
	}



/****************************************************************************
 * get_line
 ****************************************************************************/
static char *
get_line(
	char				*line,
	int				size,
	const char			*name,
	const char			*info)

	{
	const static char		space[] = "                ";
	int				i = 0;

	while ((name[i] != '\0') && (space[i] != '\0')) i++;
	if ((string_length(info) == 0) || (verbose_level < 0)) do_snprintf(line, size, "%s", name);
	else do_snprintf(line, size, "%s%s %s", name, &space[i], info);
	return (line);
	}



/****************************************************************************
 * cmd_initialize
 ****************************************************************************/
static void
cmd_initialize(
	void)

	{
	drive_init_all_devices();
	}



/****************************************************************************
 * cmd_list
 ****************************************************************************/
static void
cmd_list(
	void)

	{
	struct disk			*dsk;
	struct drive			*drv;
	char				line[3 * CWTOOL_MAX_PATH_LEN];
	int				i;

	if (disk_get(0) != NULL) printf("Available disk names are:\n");
	for (i = 0; (dsk = disk_get(i)) != NULL; i++) printf("%s\n", get_line(line, sizeof (line), disk_get_name(dsk), disk_get_info(dsk)));
	if (drive_get(0) != NULL) printf("Configured device paths are:\n");
	for (i = 0; (drv = drive_get(i)) != NULL; i++) printf("%s\n", get_line(line, sizeof (line), drive_get_path(drv), drive_get_info(drv)));
	}



/****************************************************************************
 * cmd_statistics
 ****************************************************************************/
static void
cmd_statistics(
	struct disk			*dsk)

	{
	disk_statistics(dsk, files[0]);
	}



/****************************************************************************
 * info_init
 ****************************************************************************/
static void
info_init(
	void)

	{
	info_fh = fdopen(STDERR_FILENO, "a");
	if (info_fh == NULL) error_perror2("could not reopen stderr");
	setvbuf(info_fh, NULL, _IONBF, 0);
	}



/****************************************************************************
 * info_error_sector
 ****************************************************************************/
static int
info_error_sector(
	char				*line,
	int				max,
	struct disk_sector_info		*dsk_sct_nfo,
	int				sector)

	{
	char				*reason = "mx";

	/*
	 * UGLY: access struct dsk_sct_nfo not here, better do this
	 *       do this in disk.c ?
	 */

	if (dsk_sct_nfo->flags == DISK_ERROR_FLAG_NOT_FOUND) reason = "nf";
	if (dsk_sct_nfo->flags == DISK_ERROR_FLAG_ENCODING)  reason = "en";
	if (dsk_sct_nfo->flags == DISK_ERROR_FLAG_ID)        reason = "id";
	if (dsk_sct_nfo->flags == DISK_ERROR_FLAG_NUMBERING) reason = "nu";
	if (dsk_sct_nfo->flags == DISK_ERROR_FLAG_SIZE)      reason = "si";
	if (dsk_sct_nfo->flags == DISK_ERROR_FLAG_CHECKSUM)  reason = "cs";
	return (do_snprintf(line, max, " %02d=%s@0x%06x", sector, reason, dsk_sct_nfo->offset));
	}



/****************************************************************************
 * info_error_details
 ****************************************************************************/
static void
info_error_details(
	struct disk_info		*dsk_nfo)

	{
	char				line[1024];
	int				i, l, s, t;

	verbose(-2, "details for bad sectors:");
	for (t = 0; t < CWTOOL_MAX_TRACK; t++)
		{
		for (i = s = 0; s < CWTOOL_MAX_SECTOR; s++) if (dsk_nfo->sct_nfo[t][s].flags) i++;
		if (i == 0) continue;
		for (s = 0; s < CWTOOL_MAX_SECTOR; )
			{
			l = do_snprintf(line, sizeof (line), "track %3d:", t);
			for (i = 0; (i < 4) && (s < CWTOOL_MAX_SECTOR); s++)
				{
				if (! dsk_nfo->sct_nfo[t][s].flags) continue;
				l += info_error_sector(&line[l], sizeof (line) - l, &dsk_nfo->sct_nfo[t][s], s);
				i++;
				}
			if (i > 0) verbose(-2, line);
			}
		}
	}



/****************************************************************************
 * info_print
 ****************************************************************************/
static void
info_print(
	struct disk_info		*dsk_nfo,
	int				summary)

	{
	const static char		space[] = "                                                                               ";
	char				line[1024], end = '\r', path[16];
	int				selector = 0, i = 0;

	if (verbose_level < 0) return;

	if (mode == CWTOOL_MODE_WRITE) selector += 4;
	if (summary)
		{
		end = '\n';
		selector += 2;
		if (dsk_nfo->sum.sectors_good + dsk_nfo->sum.sectors_weak + dsk_nfo->sum.sectors_bad > 0) selector++;
		}
	else if (dsk_nfo->sectors_good + dsk_nfo->sectors_weak + dsk_nfo->sectors_bad > 0) selector++;

	if (selector == 0) do_snprintf(line, sizeof (line), "reading track %3d try %2d (sectors: none) (%s)",
		dsk_nfo->track, dsk_nfo->try, string_dot(path, sizeof (path), dsk_nfo->path));
	if (selector == 1) do_snprintf(line, sizeof (line), "reading track %3d try %2d (sectors: good %2d weak %2d bad %2d) (%s)",
		dsk_nfo->track, dsk_nfo->try, dsk_nfo->sectors_good,
		dsk_nfo->sectors_weak, dsk_nfo->sectors_bad,
		string_dot(path, sizeof (path), dsk_nfo->path));
	if (selector == 2) do_snprintf(line, sizeof (line), "%3d tracks read",
		dsk_nfo->sum.tracks);
	if (selector == 3) do_snprintf(line, sizeof (line), "%3d tracks read (sectors: good %4d weak %4d bad %4d)",
		dsk_nfo->sum.tracks, dsk_nfo->sum.sectors_good,
		dsk_nfo->sum.sectors_weak, dsk_nfo->sum.sectors_bad);
	if (selector == 4) do_snprintf(line, sizeof (line), "writing track %3d (sectors: none)",
		dsk_nfo->track);
	if (selector == 5) do_snprintf(line, sizeof (line), "writing track %3d (sectors: %2d)",
		dsk_nfo->track, dsk_nfo->sectors_good);
	if (selector == 6) do_snprintf(line, sizeof (line), "%3d tracks written",
		dsk_nfo->sum.tracks);
	if (selector == 7) do_snprintf(line, sizeof (line), "%3d tracks written (sectors: %4d)",
		dsk_nfo->sum.tracks, dsk_nfo->sum.sectors_good);

	while ((line[i] != '\0') && (space[i] != '\0')) i++;
	if ((verbose_level != 0) || (debug_level != 0)) verbose(0, "%s", line);
	else fprintf(info_fh, "%s%s%c", line, &space[i], end);

	if ((summary) && (dsk_nfo->sum.sectors_bad > 0)) info_error_details(dsk_nfo);
	}



/****************************************************************************
 * cmd_read
 ****************************************************************************/
static void
cmd_read(
	struct disk			*dsk)

	{
	struct disk_option		dsk_opt = DISK_OPTION_INIT(info_print, retry, DISK_OPTION_FLAG_NONE);

	disk_read(dsk, &dsk_opt, files, files_count - 1, files[files_count - 1]);
	}



/****************************************************************************
 * cmd_write
 ****************************************************************************/
static void
cmd_write(
	struct disk			*dsk)

	{
	int				flags = (ignore_size) ? DISK_OPTION_FLAG_IGNORE_SIZE : DISK_OPTION_FLAG_NONE;
	struct disk_option		dsk_opt = DISK_OPTION_INIT(info_print, 0, flags);

	disk_write(dsk, &dsk_opt, files[0], files[1]);
	}



/****************************************************************************
 * cmdline_fill_space
 ****************************************************************************/
static void
cmdline_fill_space(
	char				*string)

	{
	while (*string != '\0') *string++ = ' ';
	}



/****************************************************************************
 * cmdline_usage
 ****************************************************************************/
static void
cmdline_usage(
	void)

	{
	char				space1[] = VERSION_STRING;
	char				space2[] = PROGRAM_NAME;

	cmdline_fill_space(space1);
	cmdline_fill_space(space2);
	printf(VERSION_STRING " | written by Karsten Scheibler\n"
		"%s | http://unusedino.de/cw/\n"
		"%s | cw@unusedino.de\n\n"
		"Usage: %s -V\n"
		"or:    %s -D\n"
		"or:    %s -I [-v] [-n] [-f <file>] [-e <config>]\n"
		"or:    %s -L [-v] [-n] [-f <file>] [-e <config>]\n"
		"or:    %s -S [-v] [-n] [-f <file>] [-e <config>]\n"
		"       %s    [--] <diskname> <srcfile>\n"
		"or:    %s -R [-v] [-n] [-f <file>] [-e <config>] [-r <num>]\n"
		"       %s    [--] <diskname> <srcfile> [<srcfile> ... ] <dstfile>\n"
		"or:    %s -W [-v] [-n] [-f <file>] [-e <config>] [-s]\n"
		"       %s    [--] <diskname> <srcfile> <dstfile>\n\n"
		"  -V            print out version\n"
		"  -D            dump builtin config\n"
		"  -I            initialize configured devices\n"
		"  -L            list available disk names\n"
		"  -S            print out statistics\n"
		"  -R            read disk\n"
		"  -W            write disk\n"
		"  -v            be more verbose\n"
#ifdef CWTOOL_DEBUG
		"  -d            increase debug level\n"
#endif /* CWTOOL_DEBUG */
		"  -n            do not read rc files\n"
		"  -f <file>     read additional config file\n"
		"  -e <config>   evaluate given string as config\n"
		"  -r <num>      number of retries if errors occur\n"
		"  -s            ignore size\n"
		"  -h            this help\n",
		space1, space1, program_name, program_name, program_name,
		program_name, program_name, space2, program_name, space2,
		program_name, space2);
	exit(0);
	}



/****************************************************************************
 * cmdline_check_arg
 ****************************************************************************/
static char *
cmdline_check_arg(
	char				*option,
	char				*msg,
	char				*file)

	{
	if (file == NULL) error("%s expects a %s", option, msg);
	return (file);
	}



/****************************************************************************
 * cmdline_check_stdin
 ****************************************************************************/
static char *
cmdline_check_stdin(
	char				*option,
	char				*file)

	{
	static int			stdin_count;

	cmdline_check_arg(option, "file name", file);
	if (string_equal(file, "-")) stdin_count++;
	if (stdin_count > 1) error("%s used stdin although already used before", option);
	return (file);
	}



/****************************************************************************
 * cmdline_check_stdout
 ****************************************************************************/
static char *
cmdline_check_stdout(
	char				*option,
	char				*file)

	{
	static int			stdout_count;

	cmdline_check_arg(option, "file name", file);
	if (string_equal(file, "-")) stdout_count++;
	if (stdout_count > 1) error("%s used stdout although already used before", option);
	return (file);
	}



/****************************************************************************
 * cmdline_min_params
 ****************************************************************************/
static int
cmdline_min_params(
	void)

	{
	if (mode == CWTOOL_MODE_READ)       return (3);
	if (mode == CWTOOL_MODE_WRITE)      return (3);
	if (mode == CWTOOL_MODE_STATISTICS) return (2);
	return (0);
	}



/****************************************************************************
 * cmdline_max_params
 ****************************************************************************/
static int
cmdline_max_params(
	void)

	{
	if (mode == CWTOOL_MODE_READ)       return (CWTOOL_MAX_IMAGES);
	if (mode == CWTOOL_MODE_WRITE)      return (3);
	if (mode == CWTOOL_MODE_STATISTICS) return (2);
	return (0);
	}



/****************************************************************************
 * cmdline_parse
 ****************************************************************************/
static void
cmdline_parse(
	int				argc,
	char				**argv)

	{
	char				*arg;
	int				args, ignore = 0, params = 0;

	for (args = 0, argv++; (arg = *argv++) != NULL; args++)
		{
		if ((arg[0] != '-') || (string_equal(arg, "-")) || (ignore))
			{
			if (mode == 0) goto bad_option;
			if (params >= cmdline_max_params()) error_message("too many parameters given");
			if (params >= 1)
				{
				if (files_count > 0) cmdline_check_stdin("<srcfile>", files[files_count - 1]);
				files[files_count++] = arg;
				}
			else disk_name = arg;
			params++;
			}
		else if (string_equal(arg, "--"))
			{
			ignore = 1;
			}
		else if ((string_equal(arg, "-h")) || (string_equal(arg, "--help")))
			{
			cmdline_usage();
			}
		else if (((string_equal(arg, "-V")) || (string_equal(arg, "--version"))) && (args == 0))
			{
			mode = CWTOOL_MODE_VERSION;
			}
		else if (((string_equal(arg, "-D")) || (string_equal(arg, "--dump"))) && (args == 0))
			{
			mode = CWTOOL_MODE_DUMP;
			}
		else if (((string_equal(arg, "-I")) || (string_equal(arg, "--initialize"))) && (args == 0))
			{
			verbose_level = -1;
			mode = CWTOOL_MODE_INITIALIZE;
			}
		else if (((string_equal(arg, "-L")) || (string_equal(arg, "--list"))) && (args == 0))
			{
			verbose_level = -1;
			mode = CWTOOL_MODE_LIST;
			}
		else if (((string_equal(arg, "-S")) || (string_equal(arg, "--statistics"))) && (args == 0))
			{
			verbose_level = -3;
			mode = CWTOOL_MODE_STATISTICS;
			}
		else if (((string_equal(arg, "-R")) || (string_equal(arg, "--read"))) && (args == 0))
			{
			verbose_level = -1;
			mode = CWTOOL_MODE_READ;
			}
		else if (((string_equal(arg, "-W")) || (string_equal(arg, "--write"))) && (args == 0))
			{
			verbose_level = -1;
			mode = CWTOOL_MODE_WRITE;
			}
		else if ((mode == 0) || (mode == CWTOOL_MODE_VERSION) || (mode == CWTOOL_MODE_DUMP))
			{
			goto bad_option;
			}
		else if ((string_equal(arg, "-v")) || (string_equal(arg, "--verbose")))
			{
			verbose_level++;
			}
#ifdef CWTOOL_DEBUG
		else if ((string_equal(arg, "-d")) || (string_equal(arg, "--debug")))
			{
			debug_level++;
			}
#endif /* CWTOOL_DEBUG */
		else if ((string_equal(arg, "-n")) || (string_equal(arg, "--no-rcfiles")))
			{
			no_rcfiles = 1;
			}
		else if ((string_equal(arg, "-f")) || (string_equal(arg, "--file")))
			{
			if (config_files_count >= CWTOOL_MAX_CONFIG_FILES) error_message("too many config files specified");
			config_files[config_files_count++] = cmdline_check_stdin("-f/--file", *argv++);
			config_type[config_types_count++]  = 'f';
			}
		else if ((string_equal(arg, "-e")) || (string_equal(arg, "--evaluate")))
			{
			if (config_evals_count >= CWTOOL_MAX_CONFIG_EVALS) error_message("too many -e/--evaluate options");
			config_evals[config_evals_count++] = cmdline_check_arg("-e/--evaluate", "parameter", *argv++);
			config_type[config_types_count++]  = 'e';
			}
		else if (((string_equal(arg, "-r")) || (string_equal(arg, "--retry"))) && (mode == CWTOOL_MODE_READ))
			{
			int		i = 0;

			if (*argv != NULL) i = sscanf(*argv++, "%d", &retry);
			if ((i != 1) || (retry < 0) || (retry > CWTOOL_MAX_RETRIES)) error_message("-r/--retry expects a valid number of retries");
			}
		else if (((string_equal(arg, "-s")) || (string_equal(arg, "--ignore-size"))) && (mode == CWTOOL_MODE_WRITE))
			{
			ignore_size = 1;
			}
		else
			{
		bad_option:
			error("unrecognized option '%s'", arg);
			}
		}
	if ((params < cmdline_min_params()) || (mode == 0)) error_message("too few parameters given");
	if (params >= 2) cmdline_check_stdout("<dstfile>", files[files_count - 1]);
	}



/****************************************************************************
 * main
 ****************************************************************************/
int
main(
	int				argc,
	char				**argv)

	{
	struct disk			*dsk;

	setlinebuf(stderr);

	/* parse cmdline */

	cmdline_parse(argc, argv);

	/* decide what to do */

	if (mode == CWTOOL_MODE_VERSION)
		{
		cmd_version();
		goto end;
		}
	if (mode == CWTOOL_MODE_DUMP)
		{
		cmd_dump();
		goto end;
		}
	config_get();
	if (mode == CWTOOL_MODE_INITIALIZE)
		{
		cmd_initialize();
		goto end;
		}
	if (mode == CWTOOL_MODE_LIST)
		{
		cmd_list();
		goto end;
		}
	dsk = disk_search(disk_name);
	if (dsk == NULL) error("unknown disk name '%s'", disk_name);
	if (mode == CWTOOL_MODE_STATISTICS)
		{
		cmd_statistics(dsk);
		goto end;
		}
	info_init();
	if (mode == CWTOOL_MODE_READ)
		{
		cmd_read(dsk);
		goto end;
		}
	if (mode == CWTOOL_MODE_WRITE)
		{
		cmd_write(dsk);
		goto end;
		}
	debug_error();
end:
	return (0);
	}
/******************************************************** Karsten Scheibler */
